File: test/parse.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'fech'
require 'json'

class Parse

  def initialize
    @temp_dir = File.join('.','temp')
    @parsed_dir = File.join('.','ruby_parsed')
    Dir.mkdir(@parsed_dir) unless Dir.exist?(@parsed_dir)
  end

  def get_id (file_name)
    file_name.sub('.fec','')
  end

  def parse (file_name)
    id = get_id(file_name)

    if !File.exist?(@parsed_dir + '/' + id + '.json')

      filing = Fech::Filing.new(id,:download_dir => @temp_dir, :csv_parser => Fech::CsvDoctor)

      rows = []

      if filing.readable?
        file = File.open(@parsed_dir + '/' + id + '.json', 'a')

        first = true

        count = 0

        file.write "{\r\n  \"rows\": [  \r\n    "
        filing.each_row do |row|
          begin
            parsed_row = filing.parse_row?(row)
            if parsed_row

              if count != 0
                file.write ",\r\n    "
              end
              count += 1

              file.write JSON.pretty_generate(parsed_row,{
                  indent: '      '
                }).gsub('}','    }')
              # file.write parsed_row.to_json
            end
          rescue Fech::VersionError => e
            STDERR.puts e
          end
        end

        puts 'parsed ' + count.to_s + ' rows'

        file.write "\r\n  ]\r\n}\r\n"

        file.close()
      end
    else
      puts 'skipping ' + id + ' because parsed file exists'
    end
  end

  def iterate
    Dir.foreach(@temp_dir) do |file|
      if file.include?('.fec') && !file.include?('fech_')
        puts 'parsing ' + file
        parse(file)
      end
    end
  end

end

parse = Parse.new
list = parse.iterate()

File: test/parse.js
var fs = require('fs'),
    _ = require('lodash'),
    queue = require('queue-async');

var parser = require('../');

var q = queue(1);

fs.readdir(__dirname + '/temp',function (err,files) {
    files.forEach(function (filename) {
        if (filename == '.DS_Store' || filename.indexOf('.fec') === -1) {
            return;
        }

        q.defer(function (filename,cb) {
            var filingId = filename.replace('.fec','');

            fs.exists(__dirname + '/node_parsed/' + filingId + '.json',function (exists) {
                if (!exists) {
//                    console.log('parsing ' + filingId + '.fec');

                    var count = 0;

                    var file = null;

                    var firstCb = true;

                    fs.createReadStream(__dirname + '/temp/' + filingId + '.fec')
                        .pipe(parser())
                        .on('data', function(row) {
                            if (row) {
                                if (count === 0) {
                                    file = fs.createWriteStream(__dirname + '/node_parsed/' + filingId + '.json');
                                    file.write('{\r\n  \"rows\": [  \r\n    ');
                                }
                                else {
                                    file.write(',\r\n    ');
                                }

                                file.write(JSON.stringify(row,null,'      ').replace('}','    }'));
                                count++;
                            }
                        })
                        .on('error',function (e) {
                            if (firstCb &&
                                e.message != 'Row type was undefined' &&
                                e.message != 'Couldn\'t find header mapping') {
//                               console.log(e);
                                cb(null);
                                firstCb = false;
                            }
                        })
                        .on('finish',function () {
                            if (count > 0) {
                                file.write('\r\n  ]\r\n}\r\n');
                                file.end();
                            }
//                            console.log('parsed ' + count + ' rows');

                            if (firstCb) {
                                cb(null);
                                firstCb = false;
                            }
                        });
                }
                else {
 //                   console.log('skipping ' + filingId + ' because parsed file exists');

                    cb(null);
                }
            });

        },filename);

    });
});
